client/asset/eth/eth.go

```diff
@@ -672,8 +672,41 @@ func (eth *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin
 
 // Redeem sends the redemption transaction, which may contain more than one
 // redemption.
-func (*ExchangeWallet) Redeem(form *asset.RedeemForm) ([]dex.Bytes, asset.Coin, uint64, error) {
-	return nil, nil, 0, asset.ErrNotImplemented
+func (eth *ExchangeWallet) Redeem(form *asset.RedeemForm) ([]dex.Bytes, asset.Coin, uint64, error) {
+	fail := func(err error) ([]dex.Bytes, asset.Coin, uint64, error) {
+		return nil, nil, 0, err
+	}
+
+	if len(form.Redemptions) == 0 {
+		return fail(errors.New("Redeem: must be called with at least 1 redemption"))
+	}
+
+	inputs := make([]dex.Bytes, 0, len(form.Redemptions))
+	var redeemedValue uint64
+	for _, redemption := range form.Redemptions {
+		var secretHash [32]byte
+		copy(secretHash[:], redemption.Spends.SecretHash)
+		if secretHash != sha256.Sum256(redemption.Secret) {
+			return fail(fmt.Errorf("Redeem: secretHash %x != sha256(%x)", secretHash, redemption.Secret))
+		}
+		swapData, err := eth.node.swap(eth.ctx, secretHash, form.AssetVersion)
+		if err != nil {
+			return nil, nil, 0, fmt.Errorf("Redeem: error finding swap state: %w", err)
+		}
+		redeemedValue += swapData.Value
+		inputs = append(inputs, redemption.Spends.Coin.ID())
+	}
+	outputCoin := eth.createAmountCoin(redeemedValue)
+	fundsRequired := dexeth.RedeemGas(len(form.Redemptions), form.AssetVersion) * form.FeeSuggestion
+
+	// TODO: make sure the amount we locked for redemption is enough to cover the gas
+	// fees.
+	_, err := eth.node.redeem(eth.ctx, form.Redemptions, form.FeeSuggestion, form.AssetVersion)
+	if err != nil {
+		return fail(fmt.Errorf("Redeem: redeem error: %w", err))
+	}
```

Comment:
Shoud a failure here unlock the coins, or does core take care of that?

Reply:
A success should unlock the coins as well. It's actually not so easy to handle the failure, since failure will actually consume some gas as well.. and if it keeps failing then we may use more gas than we had locked in the first place. Anyways, I have a TODO in there regarding this and there will be another PR regarding locking/unlocking funds for redemption, so this can be handled there.

Reply:
If the error is a communication error I don't think the fee would be lost. I'm not sure if their could be another reason for a failure here. If the swap fails I don't think there will be an error.

A TODO is fine for now imo. Will be easier to diagnose when we start actually trading with eth.

Reply:
Ah you're right, on chain error doesn't fail there. In that case we don't need to unlock funds for failure since it will be retried anyways.

Reply:
It could be a communication error though with the internal geth node and the tx is never sent over the wire. I think that's the only type of error possible, but not sure. In that case the redeem gas stays locked. Doesn't it? I could be wrong.

Reply:
Actually not even using IPC now, so not sure how this might error.

Reply:
Yeah it shouldn't be unlocked.. those funds will need to be used on the retry.

Reply:
Or, I guess the funds are not locked yet at this point?

Reply:
They will be locked when creating the order.

Reply:
Are these locked? I see `fundsRequired` saved to `t.metaData.RedemptionFeesPaid` in trade.go, but I don't see any locking or unlocking.

Reply:
https://github.com/decred/dcrdex/pull/1289/commits/990db478bba8b774accdfd71df7b33655114a0e8#r748809759

Reply:
https://github.com/decred/dcrdex/pull/1314

Reply:
Oh, ok. Thank you.

Reply:
We can figure it out there then.